app/forms/find_product_form.rb

```diff
@@ -0,0 +1,43 @@
+class FindProductForm
+  include ActiveModel::Model
+  include ActiveModel::Attributes
+  include ActiveModel::Serialization
+  include ActiveModel::Validations::Callbacks
+
+  attribute :reference, :string
+  attribute :investigation
+
+  before_validation :tidy_reference
+
+  validates :reference, numericality: { only_integer: true, greater_than: 0, message: "Enter a PSD product record reference number" }
+  validate :must_find_a_product
+  validate :must_not_find_a_product_already_linked
+
+  def product
+    product_id = reference.presence && reference.try(:to_i)
+    return nil unless product_id && product_id.positive?
+
+    Product.find_by id: product_id
```

Comment:
This will avoid repeat calls to the database. Don't need to worry about tracking changes to `reference` as it's not changed after initialisation:
```suggestion
    return @product if @product.present?
    
    product_id = reference.presence && reference.try(:to_i)
    return nil unless product_id && product_id.positive?

    @product = Product.find_by id: product_id
```